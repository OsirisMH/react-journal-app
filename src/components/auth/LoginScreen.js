import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';

import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import { removeError, setError } from '../../actions/ui';

export const LoginScreen = () => {

    const dispatch = useDispatch();

    const [ formValues, handleInputChange ] = useForm({
        email: '',
        password: ''
    });

    const { msgError, loading } = useSelector( state => state.ui );
    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if ( isFormValid() ){
            dispatch( startLoginEmailPassword(email, password) );
        }
    };

    const handleGoogleLogin = () => {
        dispatch( startGoogleLogin() );
    };

    const isFormValid = () => {
        if( email.trim().length === 0 ){
            dispatch( setError('Email is required') );
            return false;
        }
        else if ( !validator.isEmail( email ) ){
            dispatch( setError('Email is not valid') );
            return false;
        }
        else if ( password.trim().length === 0 ){
            dispatch( setError('Password is required') );
            return false;
        }

        dispatch( removeError() );
        return true;
    };

    return (
        <>
            <h3 className="auth__title mb-5">Login</h3>
        
            {
                msgError &&
                    <div className='auth__alert-error txt-center'>
                        { msgError }
                    </div>
            }

            <form 
                onSubmit={ handleLogin }
                className="animate__animated animate__fadeIn animate__faster"
            >
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange={ handleInputChange }
                />    

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    autoComplete="off"
                    value={ password }
                    onChange={ handleInputChange }
                />    

                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled = { loading }
                >
                    Login
                </button>

                <div
                    className="auth__social-networks"
                    onClick={ loading ? null : handleGoogleLogin }
                >
                    <p>Login with social networks</p>
                    <div 
                        className={"google-btn" + (loading ? ' disabled' : '')}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link
                    to="/auth/register"
                    className="link center"
                >
                    Create new account
                </Link>
            </form>  
        </>
    )
}
