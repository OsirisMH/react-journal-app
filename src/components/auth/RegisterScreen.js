import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';

import { useForm } from '../../hooks/useForm';
import { setError, removeError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {
    
    const dispatch = useDispatch()
    const { msgError } = useSelector( state => state.ui );

    const [ formValues, handleInputChange ] = useForm({
        name: '',
        email: '',
        password: '',
        confirmPassword: ''
    });

    const { name, email, password, confirmPassword } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        
        if ( isFormValid() ){
            dispatch( startRegisterWithEmailPasswordName(email, password, name) );
        }

    };  
    
    const isFormValid = () => {
        if( name.trim().length === 0 ){
            dispatch( setError('Name is required') );
            return false;
        }
        else if ( !validator.isEmail( email ) ){
            dispatch( setError('Email is not valid') );
            return false;
        }
        else if( password !== confirmPassword ){
            dispatch( setError("Password does not match") );
            return false;
        }
        else if ( password.length <= 5 ){
            dispatch( setError("Password should be at least 6 characters") );
            return false;
        }

        dispatch( removeError() );
        return true;
    };

    return (
        <>
            <h3 className="auth__title mb-5">Register</h3>

            <form
                onSubmit={ handleRegister }
                className="animate__animated animate__fadeIn animate__faster"
            >

                {
                    msgError &&
                        <div className='auth__alert-error txt-center'>
                            { msgError }
                        </div>
                }

                <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    value={ name }
                    onChange={ handleInputChange }
                />

                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange={ handleInputChange }
                />    

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    autoComplete="off"
                    value={ password }
                    onChange={ handleInputChange }
                />    

                <input
                    type="password"
                    placeholder="Confirm password"
                    name="confirmPassword"
                    className="auth__input"
                    autoComplete="off"
                    value={ confirmPassword }
                    onChange={ handleInputChange }
                /> 
                
                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                >
                    Sign up
                </button>

                <Link
                    to="/auth/login"
                    className="link center"
                >
                    Already registered?
                </Link>
            </form>  
        </>
    )
}
