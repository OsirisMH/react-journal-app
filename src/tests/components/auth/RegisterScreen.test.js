import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from 'react-router';

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};
let store = mockStore( initState );

    
const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
);

describe('Pruebas en <LoginScreen />', () => {


    test('Debe mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('Debe de mostrar el error al intentar disparar el submit sin campos', () => {
        const actions = store.getActions();
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect( actions[0] ).toEqual({
            type: types.uiSetError,
            payload: 'Name is required'
        })
    });
    
    test('Debe de mostrar la caja de alerta con error', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Name is required'
            }
        };
        let store = mockStore( initState );            
        const wrapper = mount(
            <Provider store={ store }>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        ); 

        expect( wrapper.find('.auth__alert-error').exists() ).toBe( true );
        expect( wrapper.find('.auth__alert-error').text().trim() ).toBe( initState.ui.msgError );
    });
    
});
