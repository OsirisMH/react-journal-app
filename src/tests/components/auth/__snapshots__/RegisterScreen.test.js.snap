// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> Debe mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title mb-5"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="confirmPassword"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Sign up
    </button>
    <a
      className="link center"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
